### ISSUE TITLE
[help] Dynamic, isolated states

### ISSUE THREAD

**Issue by spectras** [stefanpenner/ember-state-services#33]
Hello,

I have been pointed at your module on a question on stackoverflow: [Where to store transient ui state in Ember 2.0](http://stackoverflow.com/questions/32620550/where-to-store-transient-ui-state-in-ember-2-0). The idea of making state a service really sounds good to me and I wish it was in Ember guides.

However, I do not understand how to isolate state for component instances, or if it is even possible.

- Say I have a list-view component.
- Say it contains list-view-item components.

Sharing state between those is easy. After all, that's the point of your module. The question is how I would go about isolating the state for several instances of the list-view component.

Say I have two separate `list-view`s. A model might be shown in both. Its state is `list-view`-specific though: it can be selected in one list and not in the other.

I was thinking of tying the `stateName` to my list view's `emberId` somehow, but it might not be the best approach. Is it possible, and how can I accomplish that using ember-state-services?

**Comment by stefanpenner** [comment]
I would recommend using a composite key, that describes this.

**Comment by stefanpenner** [comment]
> The idea of making state a service really sounds good to me and I wish it was in Ember guides.

Ya once we finalize this library, it may very well end up in ember itself. This is more of a playground to explore.

**Comment by spectras** [comment]
That makes sense.
I spent a few hours studying and making tests with your module, the idea is really neat.
It works amazingly well for all long-term ui-only state. I am now using it for unsaved data of all kinds.

For the specific case of short-term, local state, I thought about the composite key for a while, and the requirement to pass the first part of the key around. Then I thought, well, if I must pass a bound value to child components, why not directly pass them the state store?

So I customized the mixin so it can be mixed into a regular object, making it a state store. Basically the same thing, except it is no longer a singleton. The main component creates an instance of it, then injects it into child components (through an attribute), and eventually destroys it.
Transferring responsibility to the main component also allows it to observe its models, and teardown relevant state when it detects some models are destroyed or go out of its scope.

I believe it still holds to the initial concept and it is more suited to that use case, in which the state is very transient (state must be cleared as soon as a model is moved out of the list - if it gets moved back, any state should be reset to initial values).

At this point is seems to be out of the scope of ember-state-services but, if it's not too much to ask, would you please care sharing your thoughts on this solution for short-term state?

**Comment by stefanpenner** [comment]
I'm curious where the existing infrastructure falls down, what you describe does not appear exceptional

**Comment by spectras** [comment]
It does not.
`ember-state-services` has been refactored and switched to weakMap, so what follows is moot, be here is what led me there at that time.

I did it using the module first. This way:
* Key used for state is `[guidFor(master component), model]`.
* Master component yields its guid alongside the models.
* Template feeds the guid to the child components as an attribute.
* Child components query by `[passed guid, model]` from the state store.

Works like a charm, but disposal of unneeded state requires manual tracking of added/removed items by the master, using an array observer. So, now it was tracking items, I figured it could just have its own state store:
* Master creates a state store. Key used is `model`.
* Master yields the state store alongside the models.
* Templates injects the state store to the child components as an attribute.
* Child components query by `model` from the state store.

At the time, I found the second approach to require less code and be easier to understand.

Now the API has changed a lot, and the switch to weakMap probably made tracking of items unneeded. I will update my code to use the new version at some point. Only sure thing until then is the issues I had are no longer valid.
